### DISCUSSION TITLE
Server Side Redirect Returns ERR_HTTP_HEADERS_SENT

### DISCUSSION THREAD

**Discussion by simonuvarov** [vercel/next.js#15310] [Help]
Hi,

I'm trying to use a redirect from `getServerSideProps` and I get `Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client` in some cases. I have seen that this problem was covered in `issues` many times and the solution was to use `res.end()` like [this](https://github.com/vercel/next.js/blob/56633ed6bebf9e4a2bcb3feff2e3de3bd61d1da9/examples/with-iron-session/pages/profile-ssr.js#L35-L48). 

However, the following code fails for me when used with `Link` or `Router.push()`:

```
export const getServerSideProps = async ({ req, res }) => {
  const session = await getCurrentSession({ req });
  if (!session) {
    res.setHeader('location', '/signin');
    res.statusCode = 302;
    res.end();
    return { props: {} };
  }

  return { props: {} };
};
```

while this one works:
```
export const getServerSideProps = async ({ req, res }) => {
  res.setHeader('location', '/signin');
  res.statusCode = 302;
  res.end();
  return { props: {} };
};

```

So, my first guess was that this happens because the first code returns the same props (?) for both paths, but I tried to return `{ props: { foo: 'bar' } };`, and it gave me the same result. 

It would be super helpful if somebody could help me understand the cause of this behavior. 

I've also gone through all links at [this RFC](https://github.com/vercel/next.js/discussions/14890)

Thanks,
Simon

**Comment by simonuvarov** [comment]
This code does not work as well:
```
export const getServerSideProps = async ({ req, res }) => {
  res.setHeader('location', '/signin');
  res.statusCode = 302;
  res.end();
  return { props: {} };
  const session = await getCurrentSession({ req });
  if (!session) {
    res.setHeader('location', '/signin');
    res.statusCode = 302;
    res.end();
    return;
  }

  return { props: {} };
};
```

and to make sure that `getCurrentSession` does not touch `res`, I tried this:
```
export const getServerSideProps = async ({ req, res }) => {
  const session = false;
  if (!session) {
    res.setHeader('location', '/signin');
    res.statusCode = 302;
    res.end();
    return { props: {} };
  }
  return { props: {} };
};
```

**Comment by paambaati** [comment]
Can you check if this happens on the `canary` version?

**Reply by simonuvarov** [reply]
I have updated `next` to `v9.4.5-canary.38` and it still returns the same error.

**Comment by simonuvarov** [comment]
Seems like this code works:
```
export const getServerSideProps = async ({ req, res }) => {
  const session = false;
  if (!session) {
    res.setHeader('location', '/signin');
    res.statusCode = 302;
    res.end();
    return;
  }
  return { props: {} };
};
```

So, I assume that `return { props: {} };` is writing in `res` after `res.end()`. 

However, when I use `return`, I get the error `Your getServerSideProps function did not return an object. Did you forget to add a return?`

With `return {}` I get `SerializableError: Error serializing props returned from getServerSideProps in "/{redacted}".`

**Reply by simonuvarov** [reply]
Weird, but this code works when deployed on Vercel.
